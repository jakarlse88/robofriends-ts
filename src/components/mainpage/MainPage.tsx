import './MainPage.css';

import CardList from '../cardlist/CardList';
import ErrorBoundary from '../errorboundary/ErrorBoundary';
import Header from '../header/Header';

import * as React from 'react';
import Scroll from '../scroll/Scroll';
import SearchBox from '../searchbox/SearchBox';

export interface IRobots {
    id: number,
    name: string,
    username: string,
    email: string
};

interface IMainPageProps {
    query: string,
    robots: IRobots[],
    actions?: any,
    children?: any
};

export default class MainPage extends React.Component<IMainPageProps> {
    constructor(props: any) {
        super(props);
    }

    public filterRobots(robots: IRobots[]): IRobots[] {
        const { query } = this.props;
        const filteredRobots: IRobots[] = robots.length ?
            robots.filter(robot => {
                return robot.name.toLowerCase().includes(query.toLowerCase());
            })
            : [];
        return filteredRobots;
    }

    public onSearchChange = (event: React.SyntheticEvent<HTMLInputElement>): void => {
        this.props.actions.setQuery(event.currentTarget.value);
    }

    public render(): JSX.Element {
        const {
            robots,
        } = this.props;

        return !robots.length ?
            <h1>Loading</h1>
            : (
                <div className='tc'>
                    <Header />
                    <SearchBox
                        searchChange={(e: any) => this.onSearchChange(e)} />
                    <Scroll>
                        <ErrorBoundary>
                            <CardList robots={this.filterRobots(robots)} />
                        </ErrorBoundary>
                    </Scroll>
                </div>
            );
    }
}