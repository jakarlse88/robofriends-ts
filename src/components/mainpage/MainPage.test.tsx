import { shallow } from 'enzyme';
import MainPage from './MainPage';

import * as React from 'react';

interface IRobots {
    id: number,
    name: string,
    username: string,
    email: string
};

let wrapper: any;
beforeEach(() => {
    const mockProps = {
        query: '',
        robots: [],
    };
    wrapper = shallow(<MainPage {...mockProps}/>);
})

describe('MainPage component', () => {
    it('successfully renders the MainPage component', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('filters robots correctly', () => {
        const mockRobots: IRobots[] = [
                {
                    email: 'john@gmail.com',
                    id: 3,
                    name: 'John',
                    username: 'johnjohn',
                },
            ];

            const query: string = 'a';

        const wrapper2: any = shallow(
            <MainPage 
                robots={mockRobots}
                query={query}
            />
        );
        expect(wrapper2.instance().filterRobots([])).toEqual([]);
        expect(wrapper2.instance().filterRobots(mockRobots)).toEqual([]);
    });
})