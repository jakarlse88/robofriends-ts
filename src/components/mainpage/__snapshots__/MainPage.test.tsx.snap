// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage component successfully renders the MainPage component 1`] = `undefined`;
